File: saber/harness/pi-runner/src/tools.ts
import {
	Any,
	Array as TypeArray,
	Boolean as TypeBoolean,
	Integer as TypeInteger,
	Literal,
	Null,
	Number as TypeNumber,
	Object as TypeObject,
	Optional,
	String as TypeString,
	Union,
	Unknown,
} from "typebox";
import { defineTool, type ToolDefinition } from "@earendil-works/pi-coding-agent";

type JsonSchema = {
	type?: string | string[];
	description?: string;
	enum?: unknown[];
	properties?: Record<string, JsonSchema>;
	items?: JsonSchema;
	required?: string[];
	additionalProperties?: boolean | JsonSchema;
};

type TaskTool = {
	name: string;
	api_name?: string;
	description?: string;
	input_schema?: JsonSchema;
	parameters?: JsonSchema;
};

type TaskMcpServer = {
	name: string;
	tools?: TaskTool[];
};

type Task = {
	setup?: {
		mcp_servers?: TaskMcpServer[];
	};
};

function withDescription(schema: any, source: JsonSchema): any {
	if (source.description) {
		return { ...schema, description: source.description };
	}
	return schema;
}

function literalSchema(value: unknown): any {
	if (value === null) return Null();
	if (["string", "number", "boolean"].includes(typeof value)) return Literal(value as string | number | boolean);
	return Unknown();
}

export function jsonSchemaToTypeBox(schema: JsonSchema | undefined): any {
	if (!schema) return TypeObject({});
	if (schema.enum && schema.enum.length > 0) {
		return withDescription(Union(schema.enum.map((value) => literalSchema(value)) as any), schema);
	}

	const typeValue = Array.isArray(schema.type) ? schema.type.find((entry) => entry !== "null") : schema.type;
	switch (typeValue) {
		case "object": {
			const required = new Set(schema.required ?? []);
			const properties = Object.fromEntries(
				Object.entries(schema.properties ?? {}).map(([name, propertySchema]) => {
					const converted = jsonSchemaToTypeBox(propertySchema);
					return [name, required.has(name) ? converted : Optional(converted)];
				}),
			);
			return withDescription(TypeObject(properties, { additionalProperties: schema.additionalProperties !== false }), schema);
		}
		case "array":
			return withDescription(TypeArray(jsonSchemaToTypeBox(schema.items ?? {})), schema);
		case "string":
			return withDescription(TypeString(), schema);
		case "integer":
			return withDescription(TypeInteger(), schema);
		case "number":
			return withDescription(TypeNumber(), schema);
		case "boolean":
			return withDescription(TypeBoolean(), schema);
		case "null":
			return withDescription(Null(), schema);
		default:
			return withDescription(Any(), schema);
	}
}

export function listTaskToolNames(task: Task): string[] {
	const names: string[] = [];
	for (const server of task.setup?.mcp_servers ?? []) {
		for (const tool of server.tools ?? []) {
			names.push(tool.api_name ?? `mcp_${server.name}_${tool.name}`);
		}
	}
	return names;
}

async function callRuntimeTool(runtimeUrl: string, toolName: string, input: Record<string, unknown>): Promise<string> {
	const response = await fetch(`${runtimeUrl.replace(/\/$/, "")}/tool`, {
		method: "POST",
		headers: { "content-type": "application/json" },
		body: JSON.stringify({ tool_name: toolName, input }),
	});
	const payload = (await response.json()) as { output?: unknown; error?: unknown };
	if (!response.ok || payload.error) {
		throw new Error(String(payload.error ?? `runtime returned HTTP ${response.status}`));
	}
	return typeof payload.output === "string" ? payload.output : JSON.stringify(payload.output ?? "");
}

export function makeSaberTools(task: Task, runtimeUrl: string): ToolDefinition[] {
	const tools: ToolDefinition[] = [
		defineTool({
			name: "bash",
			label: "Bash",
			description: "Execute a shell command in the SABER sandbox.",
			executionMode: "sequential",
			parameters: TypeObject({
				command: TypeString({ description: "The shell command to execute" }),
			}),
			execute: async (_toolCallId, params) => {
				const output = await callRuntimeTool(runtimeUrl, "bash", params as Record<string, unknown>);
				return {
					content: [{ type: "text", text: output || "(no output)" }],
					details: { saberToolName: "bash", input: params },
				};
			},
		}),
	];

	for (const server of task.setup?.mcp_servers ?? []) {
		for (const tool of server.tools ?? []) {
			const toolName = tool.api_name ?? `mcp_${server.name}_${tool.name}`;
			tools.push(
				defineTool({
					name: toolName,
					label: tool.name,
					description: tool.description ?? `Call SABER MCP tool ${toolName}.`,
					executionMode: "sequential",
					parameters: jsonSchemaToTypeBox(tool.input_schema ?? tool.parameters ?? { type: "object" }),
					execute: async (_toolCallId, params) => {
						const input = params as Record<string, unknown>;
						const output = await callRuntimeTool(runtimeUrl, toolName, input);
						return {
							content: [{ type: "text", text: output || "(no output)" }],
							details: { saberToolName: toolName, input },
						};
					},
				}),
			);
		}
	}

	return tools;
}

File: saber/harness_adapters/tools.py
from __future__ import annotations

from typing import Any


def _schema_type(json_schema: dict[str, Any]) -> Any:
    typ = json_schema.get("type")
    if typ == "string":
        return str
    if typ == "integer":
        return int
    if typ == "number":
        return float
    if typ == "boolean":
        return bool
    if typ == "array":
        return list
    if typ == "object":
        return dict
    return Any


def build_pydantic_schema(name: str, json_schema: dict[str, Any]) -> type[Any]:
    try:
        from pydantic import Field, create_model
    except ImportError as exc:
        raise RuntimeError("pydantic is required for harness tool schema conversion") from exc

    properties = json_schema.get("properties", {}) or {}
    required = set(json_schema.get("required", []) or [])
    fields: dict[str, tuple[Any, Any]] = {}

    for field_name, field_schema in properties.items():
        field_type = _schema_type(field_schema)
        description = field_schema.get("description", "")
        default = ... if field_name in required else field_schema.get("default", None)
        fields[field_name] = (field_type, Field(default, description=description))

    if not fields:
        fields["input"] = (dict, Field(default_factory=dict, description="Tool input"))

    return create_model(name, **fields)


def make_langgraph_tools(runtime: Any) -> list[Any]:
    try:
        from langchain_core.tools import StructuredTool
    except ImportError as exc:
        raise RuntimeError("langchain-core is required for LangGraph tool wrapping") from exc

    BashArgs = build_pydantic_schema(
        "BashArgs",
        {
            "type": "object",
            "properties": {
                "command": {"type": "string", "description": "The shell command to execute"},
            },
            "required": ["command"],
        },
    )

    def bash(command: str) -> str:
        return runtime.execute_tool("bash", {"command": command})

    tools = [
        StructuredTool.from_function(
            func=bash,
            name="bash",
            description="Execute a shell command in the SABER Docker sandbox.",
            args_schema=BashArgs,
        )
    ]

    for tool_name, entry in runtime.mcp.registry.items():
        tool = entry["tool"]
        description = tool.get("description", f"MCP tool {tool_name}")
        input_schema = tool.get("input_schema", {"type": "object", "properties": {}})
        args_schema = build_pydantic_schema(f"{tool_name}_Args", input_schema)

        def mcp_tool(_tool_name: str = tool_name, **kwargs: Any) -> str:
            return runtime.execute_tool(_tool_name, kwargs)

        tools.append(
            StructuredTool.from_function(
                func=mcp_tool,
                name=tool_name,
                description=description,
                args_schema=args_schema,
            )
        )

    return tools
